src/coreclr/jit/llvm.cpp

```diff
@@ -874,52 +885,170 @@ llvm::Value* Llvm::buildUserFuncCall(GenTreeCall* call)
+        {
+            if ((opAndArg.operand->gtOper == GT_CNS_INT) && opAndArg.operand->IsIconHandle())
+            {
+                void* iconValue = (void*)(opAndArg.operand->AsIntCon()->IconValue());
+                const char* methodTableName = (*_getMangledSymbolName)(_thisPtr, iconValue);
+                (*_addCodeReloc)(_thisPtr, iconValue);
+                argVec.push_back(castIfNecessary(_builder.CreateLoad(castIfNecessary(getOrCreateExternalSymbol(methodTableName), Type::getInt32PtrTy(_llvmContext)->getPointerTo())), llvmFunc->getArg(argIx)->getType()));
+            }
+            else
+            {
+                argVec.push_back(genTreeAsLlvmType(opAndArg.operand, llvmFunc->getArg(argIx)->getType()));
+            }
+            argIx++;
+        }
+        // TODO-LLVM: If the block has a handler, this will need to be an invoke.  E.g. create a CallOrInvoke as per ILToLLVMImporter
+        mapGenTreeToValue(call, _builder.CreateCall(llvmFunc, llvm::ArrayRef<Value*>(argVec)));
+        if (call->gtCallMethHnd == _compiler->eeFindHelper(CORINFO_HELP_THROW))
+        {
+            _builder.CreateUnreachable();
+        }
+    }
+}
+
+void Llvm::buildCall(GenTree* node)
+{
+    GenTreeCall* call = node->AsCall();
+    if (call->gtCallType == CT_HELPER)
+    {
+        buildHelperFuncCall(call);
     }
     else if (call->gtCallType == CT_USER_FUNC && !call->IsVirtualStub() /* TODO: Virtual stub not implemented */)
     {
         buildUserFuncCall(call);
-        return;
     }
-    failFunctionCompilation();
+    else
+    {
+        failFunctionCompilation();
+    }
 }
 
 void Llvm::buildCast(GenTreeCast* cast)
 {
-    if (cast->CastToType() == TYP_BOOL && cast->TypeIs(TYP_INT) && cast->CastOp()->TypeIs(TYP_INT))
+    var_types castToType = cast->CastToType();
+    if (castToType == TYP_BOOL && cast->TypeIs(TYP_INT) && cast->CastOp()->TypeIs(TYP_INT))
     {
-        Value* intValue = _builder.CreateZExt(getGenTreeValue(cast->CastOp()), getLLVMTypeForVarType(TYP_INT));
+        Value* intValue = _builder.CreateZExt(getGenTreeValue(cast->CastOp()), getLlvmTypeForVarType(TYP_INT));
         mapGenTreeToValue(cast, intValue); // nothing to do except map the source value to the destination GenTree
     }
-    else if (cast->CastToType() == TYP_DOUBLE && cast->CastOp()->TypeIs(TYP_FLOAT))
+    else if (castToType == TYP_DOUBLE && cast->CastOp()->TypeIs(TYP_FLOAT))
     {
-        mapGenTreeToValue(cast, _builder.CreateFPCast(getGenTreeValue(cast->CastOp()), getLLVMTypeForVarType(TYP_DOUBLE)));
+        mapGenTreeToValue(cast, _builder.CreateFPCast(getGenTreeValue(cast->CastOp()), getLlvmTypeForVarType(TYP_DOUBLE)));
+    }
+    else if (cast->TypeIs(TYP_LONG) && genActualTypeIsInt(cast->CastOp()))
+    {
+        // Cast pointer to int if necessary.  TODO-LLVM: candidate for lowering?
+        Value* sourceValue = castIfNecessary(getGenTreeValue(cast->CastOp()), getLlvmTypeForVarType(cast->CastOp()->TypeGet()));
```

Comment:
A bit curious what did this code encounter that the source was lying about the type. There was one case fixed somewhat recently in the upstream that looked sort of similar.

Definitely a candidate for lowering.

Reply:
The method is https://github.com/dotnet/runtimelab/blob/9d54885c319c09817e90928e96bd88382726907d/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L1244-L1249 for which the IR starts (node 4 is the problem):

```
------------ BB01 [???..???), preds={} succs={BB02}

------------ BB02 [000..018) (return), preds={BB01} succs={}
               [000020] ------------                 IL_OFFSET void   IL offset: 0x0
               [000021] ------------                 IL_OFFSET void   IL offset: 0x1
N001 (  3,  2) [000003] ------------         t3 =    LCL_VAR   int    V01 arg1         u:1 (last use)
                                                  /--*  t3     int
N002 (  4,  4) [000004] ---------U--         t4 = *  CAST      long <- ulong <- uint
```

Reply:
`t3` maps to arg 1 which in LLVM is a pointer type, and the cast then just tries to zext that to i64 which LLVM doesn't like

Reply:
I see, these are actual pointers, not `TYP_BYREF/TYP_REF`. Unfortunate that we have to insert these casts that have no semantic meaning, though as I understand we have no choice. We'll leave this to codegen then, there is not way to even represent "typed pointers" in IR.